gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java

```diff
@@ -0,0 +1,135 @@
+
+    private static Map<Properties, TomcatConnectionPool> poolMap =
+            new ConcurrentHashMap<Properties, TomcatConnectionPool>();
+
+    static {
+        try {
+            final ClientDRDADriver driver = registeredDriver__;
+            registeredDriver__ = new ClientPoolDriver();
+            java.sql.DriverManager.registerDriver(registeredDriver__);
+            if (driver != null) {
+                java.sql.DriverManager.deregisterDriver(driver);
+            }
+        } catch (SQLException e) {
+            // A null log writer is passed, because jdbc 1 sql exceptions are
+            // automatically traced
+            exceptionsOnLoadDriver__ = ClientExceptionUtil.newSQLException(
+                    SQLState.JDBC_DRIVER_REGISTER, e);
+        }
+        // This may possibly hit the race-condition bug of java 1.1.
+        // The Configuration static clause should execute before the following line
+        // does.
+        if (ClientConfiguration.exceptionsOnLoadResources != null) {
+            exceptionsOnLoadDriver__ = Utils.accumulateSQLException(
+                    ClientConfiguration.exceptionsOnLoadResources,
+                    exceptionsOnLoadDriver__);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public boolean acceptsURL(String url) throws SQLException {
+        return (url != null && matchProtocol(url).matches());
+    }
+
+    @Override
+    protected boolean useThriftProtocol(Matcher m) {
+        String subProtocolGroup = m.group(2);
+        String protocol = m.group(1);
+        // if USE_THRIFT_AS_DEFAULT_PROP has been set explicitly then use that
+        // for default value else use the protocol string (jdbc:snappydata://
+        //   defaults to thrift while jdbc:gemfirexd:// defaults to old DRDA)
+        return subProtocolGroup == null || subProtocolGroup.length() == 0
+                ? ClientSharedUtils.isUsingThrift(
+                protocol.equalsIgnoreCase(SNAPPY_PROTOCOL))
+                : "thrift:pool:".equalsIgnoreCase(subProtocolGroup)
+                || "pool:".equalsIgnoreCase(subProtocolGroup);
+    }
+
+    @Override
+    protected Matcher matchURL(String url) {
+        return URL_PATTERN.matcher(url);
+    }
+
+    @Override
+    protected Matcher matchProtocol(String url) {
+        return PROTOCOL_PATTERN.matcher(url);
+    }
+
+    @Override
+    public Connection connect(String url, Properties properties) throws SQLException {
+
+        if (!acceptsURL(url)) {
+            return null;
+        }
+
+        properties = (properties == null) ? new Properties() : properties;
+        String clientDriverURL = url.toLowerCase().replace("pool:", "");
+        properties.setProperty(TomcatConnectionPool.PoolProps.URL.key, clientDriverURL);
+        properties.setProperty(TomcatConnectionPool.PoolProps.DRIVER_NAME.key, ClientDriver.class.getName());
+
+        TomcatConnectionPool p = poolMap.get(properties);
+        if (p != null) {
```

Comment:
poolMap is required to have maintain connection against the Keyed properties and the value will be pooled connections. However, this can be moved out from ClienPoolDriver class and moved into the TomcatConnectionPool class itself. Did this refactoring.  Here property is acting as "keyed" only.

Reply:
Btw, we already use last one for P2P connections -- see ConnectionTable.connectionPool. That one already allows providing a connection key (which can be pool properties like you have for poolMap) and will track overall size/eviction etc. Thinking about it, the commons-pool2 might be enough for our uses since we don't need most of the extra features like statement caching/proxying etc. It looks like the existing ConnectionPool could also switch to it for the same reasons which can be tracked as a separate task.

What do you think about the available options? The basic issue being cleanup and preventing unbounded growth of poolMap (e.g. if poolMap grows to 100 and each pool to 100 then we reach 10000 connections or even more and no cleanup of pools themselves).

Reply:
**1. For Cleanup:**
The tomcat jdbc pool itself having support for the cleaning up the Idle connection. If connection stays idle for specified duration then the properties minEvictableIdleTimeMillis (default;60secs) and timeBetweenEvictionRunsMillis are checks and the idle connection are removed from the pool.
For this I have tested it across the different pool( with different keys) for every pool, internally a separate eviction thread executed and cleaned up the idle connection. 
It removes all the connections which are more the value set with **minIdle**. If we set minIdle=1 it will keep only 1 idle connection in the pool.
So, what I think is we will not require a different monitoring thread, we simply have to tune this properties according to need.

**2. For Reset (autoCommit, readOnly and IsolationLevel)**
As suggested we have added the JDBC Interceptor for this. (also the scala test). Whenever the new connection is borrowed from the pool it resets the value to default and returns the connection.